src/Core/Material/BlinnPhongMaterialModel.cpp

```diff
@@ -24,6 +25,88 @@ void BlinnPhongMaterialModel::displayInfo() const {
     print( hasNormalTexture(), " Normal Texture : ", m_texNormal );
     print( hasOpacityTexture(), " Alpha Texture  : ", m_texOpacity );
 }
+Utils::Color Material::BlinnPhongMaterialModel::evalBSDF( Vector3 w_i,
+                                                          Vector3 w_o,
+                                                          Vector3 normal,
+                                                          Vector2 uv ) {
+    // diffuse lambertien component
+    Utils::Color diffuse = m_kd / M_PI;
+
+    // Blinn-Phong specular component
+    Vector3 halfway = w_i + w_o;
+    halfway.normalize();
+    Scalar specularIntensity =
+        ( m_ns + 2.0f ) / ( 2.0f * M_PI ) * std::pow( normal.dot( halfway ), m_ns );
+    Utils::Color specular = m_ks * specularIntensity;
+
+    // Combine the diffuse and specular components
+    Utils::Color bsdf = diffuse + specular;
+
+    return bsdf;
+}
+
+std::optional<std::pair<Vector3, Scalar>> BlinnPhongMaterialModel::sample( Vector3 inDir,
```

Comment:
It is unclear for me in which frame is defined the sampled direction nor its meaning ? 
The sampler `CosineWeightedSphereSampler` returns a direction in the canonic frame (same to the local frame) , then your transformation gives it in the world frame. Is it right ?

The sampler `BlinnPhongSphereSampler` samples a cosine lobe (with power `m_ns`) around the Z axis of the canonic frame. The returned direction should then be transformed to represent a sample on an oriented lobe, either around the perfectly specular direction (`reflect(inDir, normal)`) if you want to return the outgoing direction, or around the half vector `normalize(inDir+reflect(inDir, normal))` if you want to return the halfvector.
So, before doing the transformation towards world frame, you should first orient the obtained sample wrt `reflect(inDir, normal)` or `normalize(inDir+reflect(inDir, normal))`.